File: src/TB6612.cpp
#include "TB6612.h"
#include <cmath>
#include <wiringPi.h>
#include <softPwm.h>

Motor::Motor(int In1pin, int In2pin, int PWMpin, int offset, int STBYpin){
    wiringPiSetupGpio();

    In1 = In1pin;
    In2 = In2pin;
    PWM = PWMpin;
    Standby = STBYpin;
    Offset = offset;

    pinMode(In1, OUTPUT);
    pinMode(In2, OUTPUT);
    pinMode(PWM, OUTPUT);
    softPwmCreate(PWM, 1, 255);
    pinMode(Standby, OUTPUT);
}

void Motor::drive(int speed){
    digitalWrite(Standby, HIGH);
    speed = speed * Offset;
    if(speed >= 0){
        fwd(speed);
    } else{
        rev(-speed);
    }
}

void Motor::drive(int speed, int duration){
    drive(speed);
    delay(duration);
}

void Motor::fwd(int speed){
    digitalWrite(In1, HIGH);
    digitalWrite(In2, LOW);
    softPwmWrite(PWM, speed);
}

void Motor::rev(int speed){
    digitalWrite(In1, LOW);
    digitalWrite(In2, HIGH);
    softPwmWrite(PWM, speed);
}

void Motor::brake(){
    digitalWrite(In1, HIGH);
    digitalWrite(In2, HIGH);
    softPwmWrite(PWM, 0);
}

void Motor::standby(){
    digitalWrite(Standby, LOW);
}

void forward(Motor motor1, Motor motor2, int speed){
    motor1.drive(speed);
    motor2.drive(speed);
}
void forward(Motor motor1, Motor motor2){
    motor1.drive(DEFAULTSPEED);
    motor2.drive(DEFAULTSPEED);
}

void back(Motor motor1, Motor motor2, int speed)
{
	int temp = abs(speed);
	motor1.drive(-temp);
	motor2.drive(-temp);
}
void back(Motor motor1, Motor motor2)
{
	motor1.drive(-DEFAULTSPEED);
	motor2.drive(-DEFAULTSPEED);
}

void left(Motor left, Motor right, int speed)
{
	int temp = abs(speed)/2;
	left.drive(-temp);
	right.drive(temp);
	
}
void right(Motor left, Motor right, int speed)
{
	int temp = abs(speed)/2;
	left.drive(temp);
	right.drive(-temp);
	
}

void brake(Motor motor1, Motor motor2)
{
	motor1.brake();
	motor2.brake();
}

File: Python/TB6612.py
from gpiozero import *
from time import sleep

def initDriver(STBYpin):
    Standby = OutputDevice(STBYpin)
    return Standby

class Motor:
    def __init__(self, In1pin, In2pin, PWMpin, offset, Standby):
        self.In1 = OutputDevice(In1pin)
        self.In2 = OutputDevice(In2pin)
        self.PWM = PWMOutputDevice(PWMpin)
        self.Offset = offset
        self.Standby = Standby
    
    def __fwd(self, speed):
        self.In1.on()
        self.In2.off()
        self.PWM.value = speed
    
    def __rev(self, speed):
        self.In1.off()
        self.In2.on()
        self.PWM.value = speed
    
    def drive(self, speed, duration=0):
        self.Standby.on()
        speed = speed * self.Offset
        if(speed >= 0):
            self.__fwd(speed)
        else:
            self.__rev(-speed)
        sleep(duration)

    def brake(self):
        self.In1.on()
        self.In2.on()
        self.PWM.value = 0

    def standby(self):
        self.Standby.off()
    
def forward(motor1: Motor, motor2: Motor, speed=1):
    motor1.drive(speed)
    motor2.drive(speed)
def back(motor1: Motor, motor2: Motor, speed=1):
    tmp = abs(speed)
    motor1.drive(-tmp)
    motor2.drive(-tmp)

def left(left: Motor, right: Motor, speed=1):
    tmp = abs(speed)/2
    left.drive(tmp)
    right.drive(-tmp)
def right(left: Motor, right: Motor, speed=1):
    tmp = abs(speed)/2
    left.drive(-tmp)
    right.drive(tmp)

def brake(motor1: Motor, motor2: Motor):
    motor1.brake()
    motor2.brake()
